sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1637 @@
+  `queue_state::recording` state and start recording commands to the graph
+  instance.
+
+* `propList` - Optional parameter for passing properties. Properties for
+  the `command_graph` class are defined in <<graph-properties, Graph Properties>>.
+
+Returns: `true` if any queue in `recordingQueues` has its state changed from
+`queue_state::executing` to `queue_state::recording`, `false` otherwise.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the any queue in
+  `recordingQueues` is already recording to a different graph.
+
+* Throws synchronously with error code `invalid` if any of `recordingQueues`
+  is associated with a device or context that is different from the device
+  and context used on creation of the graph.
+
+|
+[source, c++]
+----
+bool end_recording()
+----
+
+|Synchronously finishes recording on all queues that are recording to the
+graph and sets their state to `queue_state::executing`.
+
+Returns: `true` if any queue recording to the graph has its state changed from
+`queue_state::recording` to `queue_state::executing`, `false` otherwise.
+
+|
+[source, c++]
+----
+bool end_recording(queue& recordingQueue)
+----
+
+|Synchronously changes the state of `recordingQueue` to the
+`queue_state::executing` state.
+
+Parameters:
+
+* `recordingQueue` - A `sycl::queue` object to change to the executing state.
+
+Returns: `true` if `recordingQueue` has its state changed from
+`queue_state::recording` to `queue_state::executing`, `false` otherwise.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if `recordingQueue` is
+  recording to a different graph.
+
+|
+[source, c++]
+----
+bool end_recording(const std::vector<queue>& recordingQueues)
+----
+
+|Synchronously changes the state of each queue in `recordingQueues` to the
+`queue_state::executing` state.
+
+Parameters:
+
+* `recordingQueues` - List of `sycl::queue` objects to change to the executing
+  state.
+
+Returns: `true` if any queue in `recordingQueues` has its state changed from
+`queue_state::recording` to `queue_state::executing`, `false` otherwise.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if any queue in
+  `recordingQueues` is recording to a different graph.
+
+|===
+
+Table {counter: tableNumber}. Member functions of the `command_graph` class (executable graph update).
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source, c++]
+----
+void
+update(const command_graph<graph_state::modifiable>& graph);
+----
+
+|Updates the executable graph node inputs & outputs from a topologically
+identical modifiable graph. A topologically identical graph is one with the
+same structure of nodes and edges, and the nodes added in the same order to
+both graphs. Equivalent nodes in topologically identical graphs each have the
+same command, targeting the same device. There is the additional limitation that
+to update an executable graph, every node in the graph must be either a kernel
+command or a host-task.
```

Comment:
I think the concept of "topologically identical" graphs is not clear from this description.  It would help if you explained the things than can be different in the two graphs.  For example, here are some things that I *think* might be allowed based on our offline conversations:

* The arguments for a recorded kernel may be different in the second graph.  For example, the graph may capture different values for the USM pointers or accessors used in the graph.
* The code that is executed in a recorded host task may be different in the second graph.

Reply:
This section has been updated in https://github.com/reble/llvm/pull/233 to clarify the allowed differences between the graphs.

Reply:
The bullet point about host tasks seems confusing, and I can't figure out what the limitation is:

> The code that is executed in host-task nodes may be different between graphs, and the update will reflect the code from `graph` host-tasks nodes. However, any changes to executable code between kernel nodes are not counted as topologically identical and is an exception with error code `invalid`.

The first sentence seems clear.  I think it is saying that the code in the host task nodes from `graph` can be different from the code in the corresponding host task nodes from `this`.  The second sentence, though, is hard to understand, and it seems possibly contradictory to the first sentence.  What is the "executable code between kernel nodes"?  Is this different from the body of the host task nodes in the graph?

I wonder if this ability to update the content of the host task nodes is even something that you want to document as a supported operation.  Is there a real use case for this, or is it just an accident of the current implementation?

Reply:
> I wonder if this ability to update the content of the host task nodes is even something that you want to document as a supported operation. Is there a real use case for this, or is it just an accident of the current implementation?

Unfortunately it's not even an accident of the current implementation, as we haven't tried to implement update() or host-tasks yet in DPC++. The language is based on how we previously implemented update() and host-tasks in ComputeCPP, although never the combination of updating host-tasks.

Due to this speculative basis, it's probably best to leave updating host-tasks undocumented for now, or as undefined behaviour.

> What is the "executable code between kernel nodes"? Is this different from the body of the host task nodes in the graph?

Yes, this is different from host-task nodes in that it is what corresponds to an underlying PI/UR kernel object. Again however, we really haven't done enough implementation investigation to inform that we can consistently detect changes here, so I'd be inclined to also leave this as undocumented or undefined behaviour.